### PR TITLE
nixos/prosody: conditionally provision required directories with StateDirectory

### PR DESCRIPTION
**Pull request by aanderse**
###### Description of changes

- using `createHome` is a bad way to provision data directories because we don't have the ability to understand what the filesystem looks in order to make the right choices
  - maybe the user has network mounted drives we don't know about
  - maybe the user requires different permissions on this directory
  - maybe the user is using btrfs subvolumes
  - maybe something else you never considered

additionally, updated the `user` and `group` documentation to reflect the actual code and not mislead the user

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
@GrahamcOfBorg test prosody prosody-mysql

File: nixos/modules/services/networking/prosody.nix [hunk 1]
```diff
@@ -839,9 +860,8 @@ in
     users.users.prosody = mkIf (cfg.user == "prosody") {
       uid = config.ids.uids.prosody;
       description = "Prosody user";
-      createHome = true;
       inherit (cfg) group;
-      home = "${cfg.dataDir}";
+      home = cfg.dataDir;
```
Comment (pennae) [inline nixos/modules/services/networking/prosody.nix:864]: is this still necessary for prosody to function?

File: nixos/modules/services/networking/prosody.nix [hunk 2]
```diff
@@ -854,28 +874,33 @@ in
...
+          ProtectHome = true;
+          ProtectHostname = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+        }
+        (mkIf (cfg.dataDir == "/var/lib/prosody") {
+          StateDirectory = "prosody";
```
Comment (pennae) [inline nixos/modules/services/networking/prosody.nix:901]:
```suggestion
          StateDirectory = mkIf (cfg.dataDir == "/var/lib/prosody") "prosody";
```
should work just as well and not require a merge/reformat

File: nixos/modules/services/networking/prosody.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/networking/prosody.nix:864]: I had the same question. Unsure, though.

File: nixos/modules/services/networking/prosody.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/networking/prosody.nix:901]: I prefer my way as it makes the conditional standout. Too many times it is easy to gloss over a file and miss things like that.

Your opinion, my opinion, either is fine.

File: nixos/modules/services/networking/prosody.nix (hunk 1 above)
Comment (pennae) [reply nixos/modules/services/networking/prosody.nix:864]: better to leave it then. :+1:

**Review by pennae — APPROVED** [review]

**Review by mweinelt — APPROVED** [review]

**Pull request merged**